src/drivers/samv7/tone_alarm/tone_alarm.cpp

```diff
@@ -512,384 +355,82 @@ ToneAlarm::stop_note()
+	// subscribe to tune_control
+	if (_tune_control_sub < 0) {
+		_tune_control_sub = orb_subscribe(ORB_ID(tune_control));
+	}
+
+	// do we have an inter-note gap to wait for?
+	if (_silence_length > 0) {
 		stop_note();
+		work_queue(HPWORK, &_work, (worker_t)&ToneAlarm::next_trampoline, this, USEC2TICK(_silence_length));
+		_silence_length = 0;
 		return;
 	}
 
-	// parse characters out of the string until we have resolved a note
-	unsigned note = 0;
-	unsigned note_length = _note_length;
-	unsigned duration;
-
-	while (note == 0) {
-		// we always need at least one character from the string
-		int c = next_char();
-
-		if (c == 0) {
-			goto tune_end;
-		}
-
-		_next++;
-
-		switch (c) {
-		case 'L':	// select note length
-			_note_length = next_number();
-
-			if (_note_length < 1) {
-				goto tune_error;
-			}
-
-			break;
-
-		case 'O':	// select octave
-			_octave = next_number();
-
-			if (_octave > 6) {
-				_octave = 6;
-			}
-
-			break;
-
-		case '<':	// decrease octave
-			if (_octave > 0) {
-				_octave--;
-			}
-
-			break;
+	// check for updates
+	bool updated = false;
+	orb_check(_tune_control_sub, &updated);
 
-		case '>':	// increase octave
-			if (_octave < 6) {
-				_octave++;
-			}
-
-			break;
-
-		case 'M':	// select inter-note gap
-			c = next_char();
-
-			if (c == 0) {
-				goto tune_error;
-			}
-
-			_next++;
-
-			switch (c) {
-			case 'N':
-				_note_mode = MODE_NORMAL;
-				break;
-
-			case 'L':
-				_note_mode = MODE_LEGATO;
-				break;
-
-			case 'S':
-				_note_mode = MODE_STACCATO;
-				break;
-
-			case 'F':
-				_repeat = false;
-				break;
-
-			case 'B':
-				_repeat = true;
-				break;
-
-			default:
-				goto tune_error;
-			}
-
-			break;
-
-		case 'P':	// pause for a note length
-			stop_note();
-			hrt_call_after(&_note_call,
-				       (hrt_abstime)rest_duration(next_number(), next_dots()),
-				       (hrt_callout)next_trampoline,
-				       this);
-			return;
-
-		case 'T': {	// change tempo
-				unsigned nt = next_number();
-
-				if ((nt >= 32) && (nt <= 255)) {
-					_tempo = nt;
-
-				} else {
-					goto tune_error;
-				}
-
-				break;
-			}
-
-		case 'N':	// play an arbitrary note
-			note = next_number();
-
-			if (note > 84) {
-				goto tune_error;
-			}
-
-			if (note == 0) {
-				// this is a rest - pause for the current note length
-				hrt_call_after(&_note_call,
-					       (hrt_abstime)rest_duration(_note_length, next_dots()),
-					       (hrt_callout)next_trampoline,
-					       this);
-				return;
-			}
-
-			break;
-
-		case 'A'...'G':	// play a note in the current octave
-			note = _note_tab[c - 'A'] + (_octave * 12) + 1;
-			c = next_char();
-
-			switch (c) {
-			case '#':	// up a semitone
-			case '+':
-				if (note < 84) {
-					note++;
-				}
+	if (updated) {
+		orb_copy(ORB_ID(tune_control), _tune_control_sub, &_tune);
+		_play_tone = _tunes.set_control(_tune) == 0;
+	}
 
-				_next++;
-				break;
+	unsigned frequency = 0, duration = 0;
```

Comment:
Just recently I saw others refactor these to 
```
unsigned frequency = 0;
unsigned duration = 0;
```

Reply:
The tone alarm in the samv7 was never completed for that architecture port. (project dead ended) It was merged to master to allow atmel to pick up support. The code has bit rotted and was causing confusion as the the "scope" of the tunes and tones relations. The refactor to samv7 was just to remove the confusion, the physical driver portion would need to be completed. So I did not invest much time in it.

Reply:
I'm glad you took that effort upon yourself in the first place :)